#Load packages
```{r eval=FALSE, include=FALSE}
suppressPackageStartupMessages({
	library(tidyverse)
	library(qs)
	library(ggplot2)
	})
set.seed(5211314)
```

#Grouping Dot plots
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0009_1_Grouping_Dotplots.qs")

if(T){
if(T){
max_num=1;size=16
colors=c("lightgrey","#ef798a","#461220")
p <- ggplot(data, aes(x = id, y = features.plot)) +
  geom_point(aes(size = pct.exp, color = avg.exp_scaled)) +
	scale_color_gradientn(colors = colors,
						  limits = c(NA,max_num), 
						  values = scales::rescale(c(0, max_num)), 
						  breaks = c(0,max_num),  
						  labels = c("0", "max"), 
						  guide = guide_colorbar(title = "Avg. exp."),
						  oob = scales::squish) +
	scale_size_continuous(range = c(1, 6),
						  limits = c(0, 100),
						  breaks = c(0, 25, 50, 75, 100))+
	labs(x=NULL,y=NULL,size = 'Pct. exp.')+
		theme_classic()+
	theme(axis.ticks=element_line(linewidth = 0.3),
  	axis.text.x = element_text(angle = 0, 
                                   face = 1, size = size, 
                                   hjust = 0.5, vjust = 0.5, 
                                   color = "black"), 
        axis.text.y = element_text(size = size, face = "italic",
        						   color = "black", family = "sans"),
  	  panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.8),
        legend.text = element_text(size = size, face = 1), 
        legend.title = element_text(size = size, face = 1),
  	  legend.key.size= unit(size, 'pt'),
        legend.position = 'right' , 
        strip.placement = "outside", 
        strip.text.x = element_text(size = size, face="bold"),
        axis.title = element_blank())
}

names(data)
if(T){
	max_num=100
colors=c("lightgrey","#5f0a87","#3083dc","#379956","#fbec04","#b23a48","#95190c")
p <- ggplot(data, aes(x = id_all, y = features.plot)) +
  geom_point(aes(size = max_pct_exp, color = max_pct_exp)) +
	scale_color_gradientn(colors = colors,
						  limits = c(NA,max_num), 
						  values = scales::rescale(c(0, max_num)), 
						  breaks = c(0,max_num),  
						  labels = c("0", "max"),  # 自定义标签为"0"和"max"
						  guide = guide_colorbar(title = "Max exp."),
						  oob = scales::squish) +
	scale_size_continuous(range = c(1, 6),
						  limits = c(0, 100),
						  breaks = c(0, 25, 50, 75, 100))+
	labs(x=NULL,y=NULL,size = 'Pct. exp.')+
		theme_classic()+
	theme(axis.ticks=element_line(linewidth = 0.3),
		  axis.text.x = element_text(angle = 0, 
                                   face = 1, size = size, 
                                   hjust = 0.5, vjust = 0.5, 
                                   color = "black"), 
		  axis.text.y = element_blank(), 
		  axis.ticks.y = element_blank(),
        #axis.text.y = element_text(size = size, face = 1,color = "black"),
  	  panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.8),
        legend.text = element_text(size = size, face = 1), 
        legend.title = element_text(size = size, face = 1),
  	  legend.key.size= unit(size, 'pt'),
        legend.position = 'right' , 
        strip.placement = "outside", 
        strip.text.x = element_text(size = size, face="bold"),
        axis.title = element_blank())
}
}
```

#C2 GOterm_top20_gsea visualization

```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0015_tfh_c2_vs_other_GOterm_top20_gsea.qs")

if(T){
p=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  #geom_text(aes(x=0.1,y=Description,label = Description),size=3.8, hjust =0)+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 8),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 8),
        axis.line = element_line(colour = 'black', linewidth =0.3),
        axis.text.x = element_text(colour = 'black', size = 8),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 8),
  	  plot.title = element_text( size = 8,          # 新增：设置主标题字体大小为8
  	  						   hjust = 0.5, margin = margin(b = 8)),  # 左对齐并增加下边距
  	  plot.title.position = "plot",  # 标题对齐整个绘图区域
  	   legend.title = element_text(size = 8),  # 新增：图例标题字体大小
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("#65B0C6","white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+   # 反转分类顺序，使条形从上到下排列
  labs(title = "GO terms enriched in C2 vs other Tfh",
       y=c(""),
  	 x=paste0("-Log10(p value)"))
}
```

#C2 gene log2FC vs. average expression dot plot
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0016_tfh_c2_vs_other_DEG_all_merged_df.qs")

range(DEG_select$avg_log2FC);range(DEG_select$RNA.c2);
mycol <- c("#d8d8d8","black")
if(T){
p <- ggplot(DEG_select, aes(x = RNA.c2, y = avg_log2FC, color = filter_display)) +
  geom_point(size = 1.2) +  # 黑色散点
	scale_color_manual(values = mycol) +  # 应用颜色方案
  labs(title = "Expression pattern of neuron related genes",
    x = "Average expression in C2",
    y = expression(Log[2]~"(C2 vs. Other) Tfh")) +
  scale_x_continuous(limits = c(0, 35), breaks = seq(0, 35, 5)) +  # 横轴刻度
  scale_y_continuous(limits = c(0, 4.5), breaks = seq(0, 4.5, 1)) + # 纵轴刻度
  theme_bw() +
	geom_vline(xintercept = 1, color = "#99919c", linetype = "solid") +  # 新x轴参考线
	geom_hline(yintercept = 0.5, color = "#99919c", linetype = "solid")+    # 新y轴参考线
  theme(
  	axis.line = element_line(linewidth = 0.3),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", linewidth = 0.3),
    plot.title = element_text(hjust = 0.5, size = 8),
    axis.title = element_text(size = 8),
    axis.text = element_text(color = "black", size = 8),  # 所有坐标轴文本统一设置
    legend.position = "none"
  )+    #最后添加标签，浮于上方
	 geom_text_repel(  # 智能标签防重叠
    aes(label = display_gene),
    size = 3,
    box.padding = 0.4,  # 标签与点的间距
    segment.color = "black",
    segment.size = 0.2,             # 新增：设置连接线粗细（默认值为0.5，可进一步调小）
    max.overlaps = 300,
    fontface = "italic" ) + # 斜体标签
  coord_flip()  # 坐标轴翻转
}
```

#C1 GOterm_top20_gsea visualization

```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0015_tfh_c1_vs_other_GOterm_top20_gsea.qs")

if(T){
p=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  #geom_text(aes(x=0.1,y=Description,label = Description),size=3.8, hjust =0)+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 14),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 14),
        axis.line = element_line(colour = 'black', linewidth =0.5),
        axis.text.x = element_text(colour = 'black', size = 14),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 14),
  	  plot.title = element_text(hjust = 0, margin = margin(b = 15)),  # 左对齐并增加下边距
  	  plot.title.position = "plot",  # 标题对齐整个绘图区域
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("#65B0C6","white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+   # 反转分类顺序，使条形从上到下排列
  labs(title = "GO terms enriched in C1 vs other Tfh",
       y=c(""),
  	 x=c("-log10(p-value)"))

p1=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  #geom_text(aes(x=0.1,y=Description,label = Description),size=3.8, hjust =0)+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 8),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 8),
        axis.line = element_line(colour = 'black', linewidth =0.3),
        axis.text.x = element_text(colour = 'black', size = 8),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 8),
  	  plot.title = element_text( size = 8,          # 新增：设置主标题字体大小为8
  	  						   hjust = 0.5, margin = margin(b = 8)),  # 左对齐并增加下边距
  	  plot.title.position = "plot",  # 标题对齐整个绘图区域
  	   legend.title = element_text(size = 8),  # 新增：图例标题字体大小
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("#65B0C6","white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+   # 反转分类顺序，使条形从上到下排列
  labs(title = "GO terms enriched in C1 vs other Tfh",
       y=c(""),
  	 x=paste0("-Log10(p value)"))
}
```

#C1 gene log2FC vs. average expression dot plot
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0016_tfh_c1_vs_other_DEG_all_merged_df.qs")

range(DEG_select$avg_log2FC);range(DEG_select$RNA.c1);
mycol <- c("#d8d8d8","black")
if(T){
p2 <- ggplot(DEG_select, aes(x = RNA.c1, y = avg_log2FC, color = filter_display)) +
  geom_point(size = 1.2) +  # 黑色散点
	scale_color_manual(values = mycol) +  # 应用颜色方案
  labs(title = "Expression pattern of neuron related genes",
    x = "Average expression in C1",
    y = expression(Log[2]~"(C1 vs. Other) Tfh")) +
  scale_x_continuous(limits = c(0, 20), breaks = seq(0, 20, 5)) +  # 横轴刻度
  scale_y_continuous(limits = c(0, 2), breaks = seq(0, 2, 0.5)) + # 纵轴刻度
  theme_bw() +
	geom_vline(xintercept = 1, color = "#99919c", linetype = "solid") +  # 新x轴参考线
	geom_hline(yintercept = 0.5, color = "#99919c", linetype = "solid")+    # 新y轴参考线
  theme(
  	axis.line = element_line(linewidth = 0.3),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", linewidth = 0.3),
    plot.title = element_text(hjust = 0.5, size = 8),
    axis.title = element_text(size = 8),
    axis.text = element_text(color = "black", size = 8),  # 所有坐标轴文本统一设置
    legend.position = "none"
  )+    #最后添加标签，浮于上方
	 geom_text_repel(  # 智能标签防重叠
    aes(label = display_gene),
    size = 3,
    box.padding = 0.2,  # 标签与点的间距
    segment.color = "black",
    segment.size = 0.2,             # 新增：设置连接线粗细（默认值为0.5，可进一步调小）
    max.overlaps = 20,
    fontface = "italic" ) + # 斜体标签
  coord_flip()  # 坐标轴翻转
}
```

#C3 GOterm_top20_gsea visualization

```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0016_tfh_c3_vs_other_GOterm_top20_gsea.qs")

if(T){
p=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 14),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 14),
        axis.line = element_line(colour = 'black', linewidth =0.5),
        axis.text.x = element_text(colour = 'black', size = 14),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 14),
  	  plot.title = element_text(hjust = 0, margin = margin(b = 15)), 
  	  plot.title.position = "plot",  
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("#65B0C6","white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+   
  labs(title = "GO terms enriched in c3 vs other Tfh",
       y=c(""),
  	 x=c("-log10(p-value)"))
}

```

#C4 GOterm_top20_gsea visualization
```{r eval=FALSE, include=FALSE}
qload(file="object/sgNTC/0016_tfh_c4_vs_other_GOterm_top20_gsea.qs")

if(T){
p=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 14),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 14),
        axis.line = element_line(colour = 'black', linewidth =0.5),
        axis.text.x = element_text(colour = 'black', size = 14),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 14),
  	  plot.title = element_text(hjust = 0, margin = margin(b = 15)), 
  	  plot.title.position = "plot", 
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+ 
  labs(title = "GO terms enriched in c4 vs other Tfh",
       y=c(""),
  	 x=c("-log10(p-value)"))
}
```

#C5 GOterm_top20_gsea visualization

```{r eval=FALSE, include=FALSE}
qload(file="object/sgNTC/0016_tfh_c5_vs_other_GOterm_top20_gsea.qs")

if(T){
p=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 14),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 14),
        axis.line = element_line(colour = 'black', linewidth =0.5),
        axis.text.x = element_text(colour = 'black', size = 14),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 14),
  	  plot.title = element_text(hjust = 0, margin = margin(b = 15)),
  	  plot.title.position = "plot",  
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("#65B0C6","white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+ 
  labs(title = "GO terms enriched in c5 vs other Tfh",
       y=c(""),
  	 x=c("-log10(p-value)"))
}
```

#ligand-receptor analysis
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0017_tfh_DRG_related_ligand_receptor.qs")

if(T){
scale_color_ligtype<- function(...){ggplot2:::manual_scale('color',
                                                           values = setNames(coldf_ligtype$mycol, coldf_ligtype$group),...)}

if(T){
p3 = ggplot(dat_lig1,aes(id,features.plot,fill = avg.exp.scaled,size =pct.exp))+
  annotate("segment", x = -Inf, xend = df_ord_lig$end_type[20], y = df_ord_lig$lig[20], yend = df_ord_lig$lig[20],
           colour = df_ord_lig$colo[20],lwd=1)+
  annotate("segment", x = -Inf, xend = df_ord_lig$end_type[26], y = df_ord_lig$lig[26], yend = df_ord_lig$lig[26],
           colour = df_ord_lig$colo[26],lwd=1)+
  annotate("segment", x = -Inf, xend = df_ord_lig$end_type[67], y = df_ord_lig$lig[67], yend = df_ord_lig$lig[67],
           colour = df_ord_lig$colo[67],lwd=1)+
	annotate("segment", x = -Inf, xend = df_ord_lig$end_type[78], y = df_ord_lig$lig[78], yend = df_ord_lig$lig[78],
           colour = df_ord_lig$colo[78],lwd=1)+
	annotate("segment", x = -Inf, xend = df_ord_lig$end_type[84], y = df_ord_lig$lig[84], yend = df_ord_lig$lig[84],
           colour = df_ord_lig$colo[84],lwd=1)+
	annotate("segment", x = -Inf, xend = df_ord_lig$end_type[97], y = df_ord_lig$lig[97], yend = df_ord_lig$lig[97],
           colour = df_ord_lig$colo[97],lwd=1)+
	annotate("segment", x = -Inf, xend = df_ord_lig$end_type[107], y = df_ord_lig$lig[107], yend = df_ord_lig$lig[107],
           colour = df_ord_lig$colo[107],lwd=1)+

  annotate("segment", x ='c4', xend = 'c4', y = -Inf, yend = df_ord_lig$lig[20],
           colour = df_ord_lig$colo[20],lwd=1)+
	annotate("segment", x ='c5', xend = 'c5', y = -Inf, yend = df_ord_lig$lig[26],
           colour = df_ord_lig$colo[26],lwd=1)+
	annotate("segment", x ='c1', xend = 'c1', y = -Inf, yend = df_ord_lig$lig[67],
           colour = df_ord_lig$colo[67],lwd=1)+
	annotate("segment", x ='c1', xend = 'c1', y = -Inf, yend = df_ord_lig$lig[78],
           colour = df_ord_lig$colo[78],lwd=1)+
	annotate("segment", x ='c1', xend = 'c1', y = -Inf, yend = df_ord_lig$lig[84],
           colour = df_ord_lig$colo[84],lwd=1)+
	annotate("segment", x ='c2', xend = 'c2', y = -Inf, yend = df_ord_lig$lig[97],
           colour = df_ord_lig$colo[97],lwd=1)+
	annotate("segment", x ='c1', xend = 'c1', y = -Inf, yend = df_ord_lig$lig[107],
           colour = df_ord_lig$colo[107],lwd=1)+
	
  geom_point(shape=21,aes(color=lig_type),stroke=1)+
  scale_color_ligtype(guide='none')+
  theme_test(base_size = 12)+
  scale_size(range = c(0,5))+
  labs(x=NULL,y=NULL,size = 'Percent of\nexpression',fill = 'Ligand/Receptor\nexpression')+
  theme(legend.position = 'top',
  	  legend.key.size= unit(8, 'pt'),
  	  legend.text = element_text(size = 10),
  	  legend.title = element_text(size = 12))+
  scale_fill_gradientn(colours = c('grey','black'))+
	 scale_x_discrete(position = "top")+
  theme(axis.text.x = element_text(size = 15,angle = 90,hjust = 1,vjust = 0.5))+
  theme(axis.text.y = element_text(face = "italic",size = 15,color =df_ord_lig$colo))+
  theme(plot.margin = margin(5,5,5,5))
}
}
```

#ligand-receptor analysis_display_partly
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0017_tfh_DRG_related_ligand_receptor_display_partly.qs")

if(T){
pl <- ggraph(mygraph, layout = "dendrogram", circular = TRUE) +
  geom_conn_bundle(data = get_con(from = from, to = to), tension = 0.5,
                   aes(colour = ifelse(type == "ligand", as.character(celltype), NA)),
                   	show.legend = FALSE) + 
  scale_edge_color_manual(values = color_use, na.value = "grey70") + 
  geom_node_point(aes(filter = leaf, x = x * 1.01, y = y * 1.01, 
  					size = 5, color = celltype, shape = type)) +
  theme_void() + coord_fixed() + 
	guides(size = "none")+
  scale_shape_manual(values = c(ligand = 19, receptor = 1)) +
  geom_node_text(aes(x = x * 1.1, y = y * 1.1, 
  				   filter = leaf, label = label, 
  				   angle = angle, hjust =  hjust,
  				   fontface="italic",
  				   color = celltype), 
  			   size = 5, alpha = 1,
  			   show.legend = FALSE) +
  scale_colour_manual(values = color_use) +
  theme(
    plot.margin = unit(c(0, 0, 0, 0), "cm"),
    legend.key.size = unit(0.5, "cm")
  ) +
  labs(x=NULL,y=NULL,shape = "Type",color = 'Celltype')+
  expand_limits(x = c(-1.4, 1.4), y = c(-1.4, 1.4))
}
```

#gene expression violin&box&dotplot chart
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0018_geneexp_violin_dot_boxplot_chart.qs")

p1=merged_df%>%
ggplot(.,aes(x=assign_cluster,y= !!sym(i)))+
  geom_violin(
    data = merged_df %>% dplyr::filter(!!sym(i) >= 0),
    aes(fill = assign_cluster),
    alpha = 1,
    color = NA,
    trim = TRUE,
    scale = "width"
  ) +
	# 箱线图（覆盖在小提琴图上）
	geom_boxplot(data = merged_df %>% dplyr::filter(!!sym(i) > 0),
  	aes(fill = assign_cluster),
  	 fill = NA,
  	notchwidth=1,
  	staplewidth=0,
    width = 0.3,
    color = "black",
    size = 0.6,
    outliers=TRUE,
    outlier.color = "black",
    outlier.size = 0.3 
    )+
	 geom_jitter(
    aes(color = "black"),
    width = 0.4,
    alpha = 1,
    size = 0.1 
    )+
  theme_classic()+xlab("")+ylab(paste0(i," exp."))+
	scale_fill_manual(values = cols) +
	scale_color_manual(values = "black") +
	theme(axis.text.x = element_text(size = 8,color = "black"),
  	  axis.text.y = element_text(size = 8, color = "black"),
  	  axis.title.x = element_text(size = 8, color = "black"),
  	  axis.title.y = element_text(size = 8, color = "black"),
  	  title = element_text(size = 8, color = "black"),
  	  legend.position = "none",
		  )
}

```

#pie chart of tfh ratio 
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0019_tfh_group_ratio_pie_chart.qs")

if(T){
color=c("#ec5e6c","#6694cf","#cb7eb6","#f39139","#91c678","#f0e444")
# 圆环图
pl=ggplot(matrix_result,aes(x = 1, y = Count, fill = names))+
  geom_bar(width = 0.3,stat = 'identity')+
  coord_polar(theta = 'y')+
  geom_text(aes(label = paste0(round(share,1),'%')),size=7,
            position = position_stack(vjust = .5))+
  labs(x = NULL, y = NULL, fill = NULL, 
       title = "")+
  scale_fill_manual(values = color) +
  theme_void() +
  theme(axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        plot.title = element_text(hjust = 0.5, color = "#666666"),
  	  legend.key.size = unit(0.75, "cm"),   # 图例图标大小（原默认0.8cm）
    legend.text = element_text(size=14),
      plot.margin = unit(rep(0,4), "cm"))+  # 图例文字字号（原默认11）
  xlim(0.5, 1.5)

}
```

#Running GSVA with Custom Gene Sets

```{r eval=FALSE, include=FALSE}
suppressPackageStartupMessages({
	library(GSVA)
library(GSEABase)
library(tidyverse)
library(ggplot2)
#install.packages("PupillometryR")
library(PupillometryR)
library(ggforce)
library(dplyr) 
})

load(file = 'object/sgNTC/combined_seurat_filtered.Rdata')

geneset <- read.gmt(gmtfile="input/gmt/Stemlike_EffectTD_differentiation_signature.gmt")
head(geneset)

if(T){
# 按 term 列分组并收集基因
geneset_list <- geneset %>%
  group_by(term) %>%
  summarise(genes = list(gene), .groups = 'drop') %>%
  pull(genes) %>%
  set_names(unique(geneset$term))


gene.expr <-  as.matrix(cells.cut[["RNA"]]$scale.data)
dim(gene.expr)
gsva.result <- GSVA::gsva(gene.expr, 
						  gset.idx.list=geneset_list,
						  kcdf='Gaussian')

gsva_long <- reshape2::melt(gsva.result)%>%
	dplyr::rename(pathway_name="Var1",
		   cell_id="Var2",
		   pathway_score="value")
scMetadata=cells.cut@meta.data%>%
	rownames_to_column(var = "cell_id")

gsva_long_meta=merge(gsva_long,scMetadata,
				by.x ="cell_id", by.y = "cell_id", all = TRUE)

names(gsva_long_meta);min(gsva_long_meta$pathway_score);max(gsva_long_meta$pathway_score)

wilcox_test_result_list=list()
for(i in levels(gsva_long_meta$pathway_name)){
gsva_long_meta01=gsva_long_meta%>%
	dplyr::filter(pathway_name%in% i)

# 执行两两比较的 Wilcoxon 秩和检验
wilcox_test_result <- pairwise.wilcox.test(gsva_long_meta01$pathway_score,
										   gsva_long_meta01$assign_cluster, 
										   p.adjust.method = "bonferroni")
wilcox_test_result_list[[i]]=wilcox_test_result

p1 <- 
    ggplot(data = gsva_long_meta01, 
           aes(x = assign_cluster, y = pathway_score, fill = assign_cluster)) +
    geom_flat_violin(position = position_nudge(x = 0.2, y = 0), alpha = 0.8) +
    geom_point(aes(y = pathway_score, color = assign_cluster), 
               position = position_jitter(width = 0.15), size = 1, alpha = 0.1) +
    geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.8) +
    labs(y = NULL, x = NULL) +  #Normalized z-score
    guides(fill = "none", color = "none") +
    scale_y_continuous(limits = c(-0.6, 0.6)) +
    scale_fill_manual(values = c("#ec5e6c","#6694cf","#cb7eb6","#f39139","#91c678","#f0e444")) +
    scale_colour_manual(values = c("#ec5e6c","#6694cf","#cb7eb6","#f39139","#91c678","#f0e444")) +
    coord_flip() +	
	labs(title=paste0(i,"\n"))+
	theme_classic()+
	theme(axis.text.x = element_text(size = 14, color = "black", family = "sans"),
  	  axis.text.y = element_text(size = 14, color = "black", family = "sans"),
		  axis.title = element_text(size = 16),
		  plot.title = element_text(hjust = 0.5, size = 16))

pdf(paste("output/sgNTC/0020_",gsub("/", "", i),"gsva_result.pdf", sep=""),width =6, height =4,compress=TRUE)
print(p1)
dev.off()
}

save(gsva.result,
	 geneset_list,
	 scMetadata,
	 wilcox_test_result_list,
	 gsva_long_meta,
	 file = 'object/sgNTC/0020_Stemlike_EffectTD_differentiation_signature_gsva_result.Rdata')
}

rm(list =ls())
load(file = 'object/sgNTC/0020_Stemlike_EffectTD_differentiation_signature_gsva_result.Rdata')

qsavem(matrix_result,
	  file="object/sgNTC/0019_tfh_group_ratio_pie_chart.qs")


load(file = "J:\\00.keti\\000_FGT\\000_scRNAseq\\20240923_LCMV_sgNTC_vs_sgVps29/object/sgNTC/0017_tfh_DRG_related_ligand_receptor.Rdata")
```

#C1-C3 reanalysis
```{r eval=FALSE, include=FALSE}
rm(list =ls());gc()

load(file = 'object/sgNTC/combined_seurat_filtered.Rdata')

cellscut_c1_c3 <- subset(cells.cut, subset = assign_cluster=="c1" | 
					 	assign_cluster=="c2" | 
					 	assign_cluster=="c3")
cellscut_c1_c3$assign_cluster=factor(
	as.character(cellscut_c1_c3$assign_cluster))
table(cellscut_c1_c3$assign_cluster)

cellscut_c1_c3 <- NormalizeData(cellscut_c1_c3 ,
						   normalization.method = "LogNormalize", scale.factor = 10000)
#高变基因
cellscut_c1_c3 <- FindVariableFeatures(cellscut_c1_c3, selection.method = "vst", nfeatures = 3000)
gene_data <- VariableFeatures(object =cellscut_c1_c3)

#归一化
cellscut_c1_c3 <- ScaleData(cellscut_c1_c3, features = gene_data)
#降维聚类
cellscut_c1_c3 <- RunPCA(cellscut_c1_c3, 
					features = VariableFeatures(object = cellscut_c1_c3))
ElbowPlot(cellscut_c1_c3)
cellscut_c1_c3 <- FindNeighbors(cellscut_c1_c3, dims = 1:10)
cellscut_c1_c3 <- FindClusters(cellscut_c1_c3, resolution = 0.5)
cellscut_c1_c3 <- RunUMAP(cellscut_c1_c3, dims = 1:3,
					 n.components=2,
					 metric = "manhattan", 
					 #euclidean, *****
					 #cosine, ***
					 #manhattan, *****
					 #hamming, 
					 #correlation, ***
					 #precomputed
					 umap.method="umap-learn",return.model = TRUE,
					 min.dist=0.2,
					 n.neighbors=50L)

cellscut_c1_c3@active.ident=cellscut_c1_c3$assign_cluster
cellscut_c1_c3$seurat_clusters=cellscut_c1_c3@active.ident
table(cellscut_c1_c3$seurat_clusters)
```

#UMAP plot
```{r eval=FALSE, include=FALSE}
cellscut_c1_c3@active.ident=as.factor(cellscut_c1_c3$assign_cluster)
color=c("#ec5e6c","#6694cf","#cb7eb6","#f39139","#91c678","#f0e444")

if(T){
p3 <- DimPlot(cellscut_c1_c3, reduction = "umap", 
              group.by = "assign_cluster", pt.size=0.8,
              cols = color,label = TRUE,
              label.size = 5,label.color = "black",
			  order = TRUE,
              label.box = FALSE,raster=FALSE)+
	theme(axis.line = element_blank(),
		  axis.ticks = element_blank(),
		  axis.text = element_blank(),
		  panel.grid = element_blank(),
		  legend.position = "none")+
	xlab(NULL)+ylab(NULL)+	
	labs(title=NULL)

pdf(file = "output/sgNTC/0021_C1-C3_merged_UMAP.pdf",width = 5, height=4,compress = TRUE)
print(p3)
dev.off()
}
if(T){
table(cellscut_c1_c3$orig.ident)
p3 <- DimPlot(cellscut_c1_c3, reduction = "umap", 
              group.by = "orig.ident", pt.size=0.8,
              cols = color,label = FALSE,
              label.size = 5,label.color = "black",
              label.box = FALSE,raster=FALSE)+
	theme(axis.line = element_blank(),
		  axis.ticks = element_blank(),
		  axis.text = element_blank(),
		  panel.grid = element_blank(),
		  legend.position = "none")+
	xlab(NULL)+ylab(NULL)+	
	labs(title=NULL)+
  facet_wrap(~ orig.ident, ncol = 1)

pdf(file = "output/sgNTC/0021_C1-C3_merged_UMAP_origident.pdf",width = 3, height=6,compress = TRUE)
print(p3)
dev.off()
}

pbmc.markers <- FindAllMarkers(cellscut_c1_c3,test.use = "wilcox",
							   only.pos = F,min.pct = 0,logfc.threshold = 0,
							   return.thresh = 1)
write.csv(pbmc.markers, 
            file ="output/sgNTC/0021_sgNTC_C1-C3_DEG_result.csv")

save(cellscut_c1_c3,
	 pbmc.markers,
	 file = 'object/sgNTC/0021_sgNTC_C1-C3_combined_seurat_filtered.Rdata')

rm(list =ls());gc()
load(file = 'object/sgNTC/0021_sgNTC_C1-C3_combined_seurat_filtered.Rdata')
```
