#Load packages
```{r eval=FALSE, include=FALSE}
suppressPackageStartupMessages({
	library(tidyverse)
	library(qs)
	library(ggplot2)
	})
set.seed(5211314)
```

#Grouping Dot plots
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0009_1_Grouping_Dotplots.qs")

if(T){
if(T){
max_num=1;size=16
colors=c("lightgrey","#ef798a","#461220")
p <- ggplot(data, aes(x = id, y = features.plot)) +
  geom_point(aes(size = pct.exp, color = avg.exp_scaled)) +
	scale_color_gradientn(colors = colors,
						  limits = c(NA,max_num), 
						  values = scales::rescale(c(0, max_num)), 
						  breaks = c(0,max_num),  
						  labels = c("0", "max"), 
						  guide = guide_colorbar(title = "Avg. exp."),
						  oob = scales::squish) +
	scale_size_continuous(range = c(1, 6),
						  limits = c(0, 100),
						  breaks = c(0, 25, 50, 75, 100))+
	labs(x=NULL,y=NULL,size = 'Pct. exp.')+
		theme_classic()+
	theme(axis.ticks=element_line(linewidth = 0.3),
  	axis.text.x = element_text(angle = 0, 
                                   face = 1, size = size, 
                                   hjust = 0.5, vjust = 0.5, 
                                   color = "black"), 
        axis.text.y = element_text(size = size, face = "italic",
        						   color = "black", family = "sans"),
  	  panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.8),
        legend.text = element_text(size = size, face = 1), 
        legend.title = element_text(size = size, face = 1),
  	  legend.key.size= unit(size, 'pt'),
        legend.position = 'right' , 
        strip.placement = "outside", 
        strip.text.x = element_text(size = size, face="bold"),
        axis.title = element_blank())
}

names(data)
if(T){
	max_num=100
colors=c("lightgrey","#5f0a87","#3083dc","#379956","#fbec04","#b23a48","#95190c")
p <- ggplot(data, aes(x = id_all, y = features.plot)) +
  geom_point(aes(size = max_pct_exp, color = max_pct_exp)) +
	scale_color_gradientn(colors = colors,
						  limits = c(NA,max_num), 
						  values = scales::rescale(c(0, max_num)), 
						  breaks = c(0,max_num),  
						  labels = c("0", "max"),  # 自定义标签为"0"和"max"
						  guide = guide_colorbar(title = "Max exp."),
						  oob = scales::squish) +
	scale_size_continuous(range = c(1, 6),
						  limits = c(0, 100),
						  breaks = c(0, 25, 50, 75, 100))+
	labs(x=NULL,y=NULL,size = 'Pct. exp.')+
		theme_classic()+
	theme(axis.ticks=element_line(linewidth = 0.3),
		  axis.text.x = element_text(angle = 0, 
                                   face = 1, size = size, 
                                   hjust = 0.5, vjust = 0.5, 
                                   color = "black"), 
		  axis.text.y = element_blank(), 
		  axis.ticks.y = element_blank(),
        #axis.text.y = element_text(size = size, face = 1,color = "black"),
  	  panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.8),
        legend.text = element_text(size = size, face = 1), 
        legend.title = element_text(size = size, face = 1),
  	  legend.key.size= unit(size, 'pt'),
        legend.position = 'right' , 
        strip.placement = "outside", 
        strip.text.x = element_text(size = size, face="bold"),
        axis.title = element_blank())
}
}
```

#C2 GOterm_top20_gsea visualization

```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0015_tfh_c2_vs_other_GOterm_top20_gsea.qs")

if(T){
p=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  #geom_text(aes(x=0.1,y=Description,label = Description),size=3.8, hjust =0)+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 8),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 8),
        axis.line = element_line(colour = 'black', linewidth =0.3),
        axis.text.x = element_text(colour = 'black', size = 8),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 8),
  	  plot.title = element_text( size = 8,          # 新增：设置主标题字体大小为8
  	  						   hjust = 0.5, margin = margin(b = 8)),  # 左对齐并增加下边距
  	  plot.title.position = "plot",  # 标题对齐整个绘图区域
  	   legend.title = element_text(size = 8),  # 新增：图例标题字体大小
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("#65B0C6","white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+   # 反转分类顺序，使条形从上到下排列
  labs(title = "GO terms enriched in C2 vs other Tfh",
       y=c(""),
  	 x=paste0("-Log10(p value)"))
}
```

#C2 gene log2FC vs. average expression dot plot
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0016_tfh_c2_vs_other_DEG_all_merged_df.qs")

range(DEG_select$avg_log2FC);range(DEG_select$RNA.c2);
mycol <- c("#d8d8d8","black")
if(T){
p <- ggplot(DEG_select, aes(x = RNA.c2, y = avg_log2FC, color = filter_display)) +
  geom_point(size = 1.2) +  # 黑色散点
	scale_color_manual(values = mycol) +  # 应用颜色方案
  labs(title = "Expression pattern of neuron related genes",
    x = "Average expression in C2",
    y = expression(Log[2]~"(C2 vs. Other) Tfh")) +
  scale_x_continuous(limits = c(0, 35), breaks = seq(0, 35, 5)) +  # 横轴刻度
  scale_y_continuous(limits = c(0, 4.5), breaks = seq(0, 4.5, 1)) + # 纵轴刻度
  theme_bw() +
	geom_vline(xintercept = 1, color = "#99919c", linetype = "solid") +  # 新x轴参考线
	geom_hline(yintercept = 0.5, color = "#99919c", linetype = "solid")+    # 新y轴参考线
  theme(
  	axis.line = element_line(linewidth = 0.3),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", linewidth = 0.3),
    plot.title = element_text(hjust = 0.5, size = 8),
    axis.title = element_text(size = 8),
    axis.text = element_text(color = "black", size = 8),  # 所有坐标轴文本统一设置
    legend.position = "none"
  )+    #最后添加标签，浮于上方
	 geom_text_repel(  # 智能标签防重叠
    aes(label = display_gene),
    size = 3,
    box.padding = 0.4,  # 标签与点的间距
    segment.color = "black",
    segment.size = 0.2,             # 新增：设置连接线粗细（默认值为0.5，可进一步调小）
    max.overlaps = 300,
    fontface = "italic" ) + # 斜体标签
  coord_flip()  # 坐标轴翻转
}
```

#C1 GOterm_top20_gsea visualization

```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0015_tfh_c1_vs_other_GOterm_top20_gsea.qs")

if(T){
p=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  #geom_text(aes(x=0.1,y=Description,label = Description),size=3.8, hjust =0)+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 14),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 14),
        axis.line = element_line(colour = 'black', linewidth =0.5),
        axis.text.x = element_text(colour = 'black', size = 14),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 14),
  	  plot.title = element_text(hjust = 0, margin = margin(b = 15)),  # 左对齐并增加下边距
  	  plot.title.position = "plot",  # 标题对齐整个绘图区域
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("#65B0C6","white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+   # 反转分类顺序，使条形从上到下排列
  labs(title = "GO terms enriched in C1 vs other Tfh",
       y=c(""),
  	 x=c("-log10(p-value)"))

p1=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  #geom_text(aes(x=0.1,y=Description,label = Description),size=3.8, hjust =0)+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 8),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 8),
        axis.line = element_line(colour = 'black', linewidth =0.3),
        axis.text.x = element_text(colour = 'black', size = 8),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 8),
  	  plot.title = element_text( size = 8,          # 新增：设置主标题字体大小为8
  	  						   hjust = 0.5, margin = margin(b = 8)),  # 左对齐并增加下边距
  	  plot.title.position = "plot",  # 标题对齐整个绘图区域
  	   legend.title = element_text(size = 8),  # 新增：图例标题字体大小
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("#65B0C6","white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+   # 反转分类顺序，使条形从上到下排列
  labs(title = "GO terms enriched in C1 vs other Tfh",
       y=c(""),
  	 x=paste0("-Log10(p value)"))
}
```

#C1 gene log2FC vs. average expression dot plot
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0016_tfh_c1_vs_other_DEG_all_merged_df.qs")

range(DEG_select$avg_log2FC);range(DEG_select$RNA.c1);
mycol <- c("#d8d8d8","black")
if(T){
p2 <- ggplot(DEG_select, aes(x = RNA.c1, y = avg_log2FC, color = filter_display)) +
  geom_point(size = 1.2) +  # 黑色散点
	scale_color_manual(values = mycol) +  # 应用颜色方案
  labs(title = "Expression pattern of neuron related genes",
    x = "Average expression in C1",
    y = expression(Log[2]~"(C1 vs. Other) Tfh")) +
  scale_x_continuous(limits = c(0, 20), breaks = seq(0, 20, 5)) +  # 横轴刻度
  scale_y_continuous(limits = c(0, 2), breaks = seq(0, 2, 0.5)) + # 纵轴刻度
  theme_bw() +
	geom_vline(xintercept = 1, color = "#99919c", linetype = "solid") +  # 新x轴参考线
	geom_hline(yintercept = 0.5, color = "#99919c", linetype = "solid")+    # 新y轴参考线
  theme(
  	axis.line = element_line(linewidth = 0.3),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", linewidth = 0.3),
    plot.title = element_text(hjust = 0.5, size = 8),
    axis.title = element_text(size = 8),
    axis.text = element_text(color = "black", size = 8),  # 所有坐标轴文本统一设置
    legend.position = "none"
  )+    #最后添加标签，浮于上方
	 geom_text_repel(  # 智能标签防重叠
    aes(label = display_gene),
    size = 3,
    box.padding = 0.2,  # 标签与点的间距
    segment.color = "black",
    segment.size = 0.2,             # 新增：设置连接线粗细（默认值为0.5，可进一步调小）
    max.overlaps = 20,
    fontface = "italic" ) + # 斜体标签
  coord_flip()  # 坐标轴翻转
}
```

#C3 GOterm_top20_gsea visualization

```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0016_tfh_c3_vs_other_GOterm_top20_gsea.qs")

if(T){
p=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 14),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 14),
        axis.line = element_line(colour = 'black', linewidth =0.5),
        axis.text.x = element_text(colour = 'black', size = 14),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 14),
  	  plot.title = element_text(hjust = 0, margin = margin(b = 15)), 
  	  plot.title.position = "plot",  
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("#65B0C6","white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+   
  labs(title = "GO terms enriched in c3 vs other Tfh",
       y=c(""),
  	 x=c("-log10(p-value)"))
}

```

#C4 GOterm_top20_gsea visualization
```{r eval=FALSE, include=FALSE}
qload(file="object/sgNTC/0016_tfh_c4_vs_other_GOterm_top20_gsea.qs")

if(T){
p=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 14),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 14),
        axis.line = element_line(colour = 'black', linewidth =0.5),
        axis.text.x = element_text(colour = 'black', size = 14),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 14),
  	  plot.title = element_text(hjust = 0, margin = margin(b = 15)), 
  	  plot.title.position = "plot", 
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+ 
  labs(title = "GO terms enriched in c4 vs other Tfh",
       y=c(""),
  	 x=c("-log10(p-value)"))
}
```

#C5 GOterm_top20_gsea visualization

```{r eval=FALSE, include=FALSE}
qload(file="object/sgNTC/0016_tfh_c5_vs_other_GOterm_top20_gsea.qs")

if(T){
p=ggplot(gsea_results_df, aes(x = -log10(pvalue), y = Description, fill = path_group))+
  geom_bar(stat = "identity", width = 0.7, colour = "black")+
  theme_classic()+
  theme(axis.text.y = element_text(colour = 'black', size = 14),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(colour = 'black', size = 14),
        axis.line = element_line(colour = 'black', linewidth =0.5),
        axis.text.x = element_text(colour = 'black', size = 14),
        axis.ticks.x = element_line(colour = 'black'),
        axis.title.x = element_text(colour = 'black', size = 14),
  	  plot.title = element_text(hjust = 0, margin = margin(b = 15)),
  	  plot.title.position = "plot",  
        legend.position = "none")+ 
  scale_x_continuous(expand = c(0,0))+
  scale_fill_manual(values = c("#65B0C6","white"))+
  scale_y_discrete(expand = c(0.05,0),limits = rev)+ 
  labs(title = "GO terms enriched in c5 vs other Tfh",
       y=c(""),
  	 x=c("-log10(p-value)"))
}
```

#ligand-receptor analysis
```{r eval=FALSE, include=FALSE}
qload("object/sgNTC/0017_tfh_DRG_related_ligand_receptor.qs")

if(T){
scale_color_ligtype<- function(...){ggplot2:::manual_scale('color',
                                                           values = setNames(coldf_ligtype$mycol, coldf_ligtype$group),...)}

if(T){
p3 = ggplot(dat_lig1,aes(id,features.plot,fill = avg.exp.scaled,size =pct.exp))+
  annotate("segment", x = -Inf, xend = df_ord_lig$end_type[20], y = df_ord_lig$lig[20], yend = df_ord_lig$lig[20],
           colour = df_ord_lig$colo[20],lwd=1)+
  annotate("segment", x = -Inf, xend = df_ord_lig$end_type[26], y = df_ord_lig$lig[26], yend = df_ord_lig$lig[26],
           colour = df_ord_lig$colo[26],lwd=1)+
  annotate("segment", x = -Inf, xend = df_ord_lig$end_type[67], y = df_ord_lig$lig[67], yend = df_ord_lig$lig[67],
           colour = df_ord_lig$colo[67],lwd=1)+
	annotate("segment", x = -Inf, xend = df_ord_lig$end_type[78], y = df_ord_lig$lig[78], yend = df_ord_lig$lig[78],
           colour = df_ord_lig$colo[78],lwd=1)+
	annotate("segment", x = -Inf, xend = df_ord_lig$end_type[84], y = df_ord_lig$lig[84], yend = df_ord_lig$lig[84],
           colour = df_ord_lig$colo[84],lwd=1)+
	annotate("segment", x = -Inf, xend = df_ord_lig$end_type[97], y = df_ord_lig$lig[97], yend = df_ord_lig$lig[97],
           colour = df_ord_lig$colo[97],lwd=1)+
	annotate("segment", x = -Inf, xend = df_ord_lig$end_type[107], y = df_ord_lig$lig[107], yend = df_ord_lig$lig[107],
           colour = df_ord_lig$colo[107],lwd=1)+

  annotate("segment", x ='c4', xend = 'c4', y = -Inf, yend = df_ord_lig$lig[20],
           colour = df_ord_lig$colo[20],lwd=1)+
	annotate("segment", x ='c5', xend = 'c5', y = -Inf, yend = df_ord_lig$lig[26],
           colour = df_ord_lig$colo[26],lwd=1)+
	annotate("segment", x ='c1', xend = 'c1', y = -Inf, yend = df_ord_lig$lig[67],
           colour = df_ord_lig$colo[67],lwd=1)+
	annotate("segment", x ='c1', xend = 'c1', y = -Inf, yend = df_ord_lig$lig[78],
           colour = df_ord_lig$colo[78],lwd=1)+
	annotate("segment", x ='c1', xend = 'c1', y = -Inf, yend = df_ord_lig$lig[84],
           colour = df_ord_lig$colo[84],lwd=1)+
	annotate("segment", x ='c2', xend = 'c2', y = -Inf, yend = df_ord_lig$lig[97],
           colour = df_ord_lig$colo[97],lwd=1)+
	annotate("segment", x ='c1', xend = 'c1', y = -Inf, yend = df_ord_lig$lig[107],
           colour = df_ord_lig$colo[107],lwd=1)+
	
  geom_point(shape=21,aes(color=lig_type),stroke=1)+
  scale_color_ligtype(guide='none')+
  theme_test(base_size = 12)+
  scale_size(range = c(0,5))+
  labs(x=NULL,y=NULL,size = 'Percent of\nexpression',fill = 'Ligand/Receptor\nexpression')+
  theme(legend.position = 'top',
  	  legend.key.size= unit(8, 'pt'),
  	  legend.text = element_text(size = 10),
  	  legend.title = element_text(size = 12))+
  scale_fill_gradientn(colours = c('grey','black'))+
	 scale_x_discrete(position = "top")+
  theme(axis.text.x = element_text(size = 15,angle = 90,hjust = 1,vjust = 0.5))+
  theme(axis.text.y = element_text(face = "italic",size = 15,color =df_ord_lig$colo))+
  theme(plot.margin = margin(5,5,5,5))
}
}
qsavem(dat_lig1,
	 df_ord_lig,
	  file="object/sgNTC/0017_tfh_DRG_related_ligand_receptor.qs")
load(file = "J:\\00.keti\\000_FGT\\000_scRNAseq\\20240923_LCMV_sgNTC_vs_sgVps29/object/sgNTC/0017_tfh_DRG_related_ligand_receptor.Rdata")
```

#ligand-receptor analysis小图展示图
```{r eval=FALSE, include=FALSE}
suppressPackageStartupMessages({
	library(data.table)
	library(dplyr)
	library(gridExtra)
	library(grid)
	library(ggrepel)
	library(readr)
	library(ggplot2)
	library(ggpubr)
	library(ggraph)
	library(igraph)
	library(tidyverse)
	library(RColorBrewer)
	library(rstatix)
	library(grDevices)
	library(circlize)
})

load(file = "object/sgNTC/0017_tfh_DRG_related_ligand_receptor.Rdata")
gene_name1

small_legend <- function(fontsize = 5, keysize = .1, marginsize = c(-.1, 0, 0, 0), ...) {    
	small_legend_theme <- theme(legend.title = element_text(size = fontsize), 
								legend.text = element_text(size = fontsize),
								legend.key.size = unit(keysize, "lines"),
								legend.margin = margin(marginsize[1], 
													   marginsize[2], 
													   marginsize[3], 
													   marginsize[4], 
													   unit = "cm"),
								...
								)    
	return(small_legend_theme)
}

colorF <- c("#FDB462","#87638F", "#84cbda","#FB8072","#8DD3C7", "#BEBADA", "#80B1D3",
			 "#80B1D3", "#ed995a", "#FB8072", "#f3c89d", "#8cc598", "#3d8ea9")

if(T){
df=data.frame(
	"source"=c("DRG_neurons","DRG_neurons","Tfh_c2","Tfh_c2",
			   "Tfh_c1","Tfh_c1","Tfh_c1","Tfh_c2",
			   "DRG_neurons","DRG_neurons","DRG_neurons"),
	"target"=c("Tfh_c1","Tfh_c5","DRG_neurons","DRG_neurons",
			   "DRG_neurons","DRG_neurons","DRG_neurons","DRG_neurons",
			   "Tfh_c4","Tfh_c5","Tfh_c4"),
	"ligand"=c("Vcam1","Vcam1","Icam1","Icam1",
			   "Btla","Cd40lg","Il21","Il21",
			   "Icam1","Icam1","Icam1"),
	"receptor"=c("Itgb1","Itgb1","Itgb2","Itgal",
				 "Tnfrsf14","Itgb2","Il21r","Il21r",
				 "Itgb2","Itgal","Itgal"),
  stringsAsFactors = FALSE # 确保初始时不自动转换为因子
)

# 将指定的列转换为因子
df$source <- factor(df$source)
df$target <- factor(df$target)
df$ligand <- factor(df$ligand)
df$receptor <- factor(df$receptor)
}

if(T){
plotData <- df %>%
    dplyr::arrange(target, ligand, receptor)
color_use <- c("#3d8ea9","#ec5e6c","#39268f","#f39139","#91c678","#87638F")

sep <- ">@<"
vertices <- data.frame(name = c("root", unique(plotData$source), unique(plotData$target)), 
					   type = NA, celltype = NA, label = NA)
for (i in c(unique(plotData$source))) {
    vertices <- rbind(vertices, data.frame(name = paste0(i, sep, "ligand"), 
    									   type = NA, celltype = i, label = NA))
}
for (i in c(unique(plotData$target))) {
    vertices <- rbind(vertices, data.frame(name = paste0(i, sep, "receptor"), type = NA, celltype = i, label = NA))
}
vertices2 <- data.frame()
for (i in seq_len(nrow(plotData))) {
    vertices2 <- rbind(vertices2, data.frame(
        name = paste0(plotData$source[i], sep, plotData$ligand[i]),
        type = "ligand", celltype = plotData$source[i], label = plotData$ligand[i]
    ))
}
for (i in seq_len(nrow(plotData))) {
    vertices2 <- rbind(vertices2, data.frame(
        name = paste0(plotData$target[i], sep, plotData$receptor[i]),
        type = "receptor", celltype = plotData$target[i], label = plotData$receptor[i]
    ))
}
vertices2 <- vertices2 %>% dplyr::arrange(celltype)
vertices <- rbind(vertices, vertices2)
vertices <- vertices[!duplicated(vertices, fromLast = T), ]


edges <- data.frame(
    from = c(rep("root", length(c(unique(plotData$source), unique(plotData$target)))), unique(plotData$source), unique(plotData$target)),
    to = c(
        unique(plotData$source), unique(plotData$target), paste0(unique(plotData$source), sep, "ligand"),
        paste0(unique(plotData$target), sep, "receptor")
    ), name = NA, edgeColor = NA
)

edges2 <- data.frame()
for (i in seq_len(nrow(plotData))) {
    edges2 <- rbind(edges2, data.frame(
        from = paste0(plotData$source[i], sep, "ligand"),
        to = paste0(plotData$source[i], sep, plotData$ligand[i]),
        name = paste0(plotData$source[i], sep, plotData$ligand[i], sep, plotData$target[i], sep, plotData$receptor[i]),
        edgeColor = plotData$target[i]
    ))
}

for (i in seq_len(nrow(plotData))) {
    edges2 <- rbind(edges2, data.frame(
        from = paste0(plotData$target[i], sep, "receptor"),
        to = paste0(plotData$target[i], sep, plotData$receptor[i]),
        name = paste0(plotData$source[i], sep, plotData$ligand[i], sep, plotData$target[i], sep, plotData$receptor[i]),
        edgeColor = plotData$target[i]
    ))
}
edges2 <- edges2 %>% dplyr::arrange(edgeColor)
edges <- rbind(edges, edges2)
edges <- edges[!duplicated(edges, fromLast = T), ]

from <- match(paste0(plotData$source, sep, plotData$ligand), vertices$name)
to <- match(paste0(plotData$target, sep, plotData$receptor), vertices$name)

vertices$id <- NA
myleaves <- which(is.na(match(vertices$name, edges$from)))
nleaves <- length(myleaves)
vertices$id[myleaves] <- seq(1:nleaves)
vertices$angle <- 90 - 360 * vertices$id / nleaves
vertices$hjust <- ifelse(vertices$angle < -90, 1, 0)
vertices$angle <- ifelse(vertices$angle < -90, vertices$angle + 180, vertices$angle)

mygraph <- igraph::graph_from_data_frame(edges, vertices = vertices, directed = TRUE)

if(T){
pl <- ggraph(mygraph, layout = "dendrogram", circular = TRUE) +
  geom_conn_bundle(data = get_con(from = from, to = to), tension = 0.5,
                   aes(colour = ifelse(type == "ligand", as.character(celltype), NA)),
                   	show.legend = FALSE) + 
  scale_edge_color_manual(values = color_use, na.value = "grey70") + # 定义颜色方案，并为非 ligand 类型设置默认颜色
  geom_node_point(aes(filter = leaf, x = x * 1.01, y = y * 1.01, 
  					size = 5, color = celltype, shape = type)) +
  theme_void() + coord_fixed() + 
	guides(size = "none")+
  scale_shape_manual(values = c(ligand = 19, receptor = 1)) +
  geom_node_text(aes(x = x * 1.1, y = y * 1.1, 
  				   filter = leaf, label = label, 
  				   angle = angle, hjust =  hjust,
  				   #angle = 0, hjust = 0.5,vjust=-1.5,
  				   fontface="italic",
  				   color = celltype), 
  			   size = 5, alpha = 1,
  			   show.legend = FALSE) +
  scale_colour_manual(values = color_use) + # 为节点和文本定义颜色方案
  theme(
    plot.margin = unit(c(0, 0, 0, 0), "cm"),
    legend.key.size = unit(0.5, "cm") # 修改图例大小
  ) +
  labs(x=NULL,y=NULL,shape = "Type",color = 'Celltype')+
  expand_limits(x = c(-1.4, 1.4), y = c(-1.4, 1.4))

pdf("output/sgNTC/0017_tfh_DRG_related_ligand_receptor_display_partly.pdf", w=5, h=4)
print(pl)
dev.off()

graph2ppt(pl,"output/sgNTC/0017_tfh_DRG_related_ligand_receptor_display_partly.pptx",
		  width =5, height =4)
}
}

save(df,
	 vertices,
	 mygraph,
	 pl,
       file = "object/sgNTC/0017_tfh_DRG_related_ligand_receptor_display_partly.Rdata")

load(file = "object/sgNTC/0017_tfh_DRG_related_ligand_receptor_display_partly.Rdata")
```

#gene expression VlnPlot & UMAP
```{r eval=FALSE, include=FALSE}
load(file = 'object/sgNTC/combined_seurat_filtered.Rdata')
##提取标准化表达矩阵
mat=as.data.frame(GetAssayData(cells.cut,layer = 'data'))%>%
	rownames_to_column(var = "rowname")%>%
	column_to_rownames(var = "rowname")
dim(mat);mat[1:10,1:10]
table(mat["Cd69",]>2.5)


head(a)
color=c("#ec5e6c","#6694cf","#cb7eb6","#f39139","#91c678","#f0e444")

pdf("output/sgNTC/0018_tfh_gene_expression_vlnplot.pdf", w=10, h=20,compress=TRUE)
VlnPlot(cells.cut, features = c("Bcl6","Cxcr5","Pdcd1","Tbx21","Tcf7","Il21","Il4","Ifng"),
		pt.size =0.3,cols = color,
		ncol = 2)
dev.off()

pdf("output/sgNTC/0018_tfh_gene_expression_vlnplot_02.pdf", w=10, h=20,compress=TRUE)
VlnPlot(cells.cut, features = c("Itgb1","Icam1","Btla","Cd40lg","Il21","Itgal","Itgb2"),
		pt.size =0.3,cols = color,
		ncol = 2)
dev.off()

if(T){
features=c("tox","tox2","il2","nt5e",
		   "cd69","egr1","jun","fosb","fos",
		   "junb","nr4a1","cxcl10","il21r",
		   "id3","ccr7","slamf6","tnfrsf4",
		   "izumo1r","ramp3","Klrg1","CXCR6",
		   "CX3CR1","cxcr3","gzma","gzmb","ccr2",
		   "prf1","ccl3","ccl4","ccl5","Havcr2",
		   "Entpd1","il2ra",
		   "prdm1","il7r","sell","bcl2",
		   "ccr5","id2","mt1","mt2","myb",
		   "myc","tnfrsf9","mki67","cks2",
		   "mad2l1","ccnb1","xcl1","gzmk",
		   "gzmm","hells","gnl3","batf",
		   "lag3","cd160","ctla4","tnfrsf14",
		   "icos","cd28","cd226","fasl","lef1",
		   "Bcl6","Cxcr5","Pdcd1","Tbx21","Tcf7",
		   "Il21","Il4","Ifng","Itgb1","Icam1",
		   "Btla","Cd40lg","Il21","Itgal","Itgb2")

# Step 1: 全小写
features <- tolower(features)

# Step 2: 首字母大写（任选一种方法）
# 方法一：基础R
features <- paste0(toupper(substr(features, 1, 1)), substr(features, 2, nchar(features)))
# 方法二：正则表达式
features <- sub("^(.)", "\\U\\1", features, perl = TRUE)
# 方法三：stringr包
features <- stringr::str_to_title(features)
}

if(T){
pdf("output/sgNTC/0018_tfh_gene_expression_vlnplot_03.pdf", w=15, h=60,compress=TRUE)
VlnPlot(cells.cut, features = features,
		pt.size =0.3,cols = color,
		ncol = 3)
dev.off()
}

if(T){
	max_num=3.3
colors=c("lightgrey","#5f0a87","#3083dc","#379956","#fbec04","#b23a48","#95190c")
p <- FeaturePlot(cells.cut, 
                   features = 'Cd69',
                   pt.size = 0.5,
                   order = TRUE) +
   scale_color_gradientn(
    colors = colors,
    limits = c(NA,max_num),  # 设置最大值上限为3，最小值自动适配
    values = scales::rescale(c(0, max_num)),  # 重新映射颜色范围到[0,3]
      breaks = c(0,max_num),  # 指定图例的断点为0和最大值
    labels = c("0", "max"),  # 自定义标签为"0"和"max"
    oob = scales::squish  )+  # 将超出范围的数值压缩到颜色标度内
    theme(
      axis.line = element_blank(),
      axis.ticks = element_blank(),
      axis.text = element_blank(),
      panel.grid = element_blank(),
      plot.margin = unit(rep(0,4), "cm"), 
      legend.position = "left"
    ) +
    xlab(NULL) + ylab(NULL)
pdf(file = "output/sgNTC/0018_tfh_gene_expression_UMAP_legend.pdf",
	width = 5, height=4.5,compress = TRUE)
print(p)
dev.off()
}

colors=c("lightgrey","#5f0a87","#3083dc","#379956","#fbec04","#b23a48","#95190c")
if(T){
if(!dir.exists("output/sgNTC/gene_expression_UMAP")) dir.create(
	"output/sgNTC/gene_expression_UMAP", recursive = TRUE)

# 循环生成所有基因的FeaturePlot
for(gene in features){
		# 检查目的基因行中大于2.5的元素数量
aim_mat=mat[gene, ]
	count_gt_1 <- sum(aim_mat > 1, na.rm = TRUE)

# 条件判断与分位数计算
if (count_gt_1 > ncol(mat)/20) {
  # 计算1/20分位数（即第5%分位数）
  max_num <- quantile(aim_mat, 
                     probs = 1 - 1/20, 
                     na.rm = TRUE,
                     type = 7)  # 使用默认分位数计算方法
  max_num <- unname(max_num)   # 移除分位数名称
} else {
  max_num <- 1  # 或其他默认值
}
  p <- FeaturePlot(cells.cut, 
                   features = gene,
                   pt.size = 0.5,
                   order = TRUE) +
   scale_color_gradientn(
    colors = colors,
    limits = c(NA, max_num),  # 设置最大值上限为3，最小值自动适配
    values = scales::rescale(c(0, max_num)),  # 重新映射颜色范围到[0,3]
    oob = scales::squish  )+  # 将超出范围的数值压缩到颜色标度内
    theme(
      axis.line = element_blank(),
      axis.ticks = element_blank(),
      axis.text = element_blank(),
      panel.grid = element_blank(),
      plot.margin = unit(rep(0,4), "cm"), 
      legend.position = "none"
    ) +
    xlab(NULL) + ylab(NULL)
  print(paste0("now make UMAP chart of ",gene))
  # 构建动态文件名
  pdf_file <- paste0("output/sgNTC/gene_expression_UMAP/tfh_gene_expression_UMAP_", gene, ".pdf")
  
  pdf(file = pdf_file, width = 5, height = 4.5, compress = TRUE)
  print(p)
  dev.off()
  }}
```

#gene expression violin&box&dotplot chart
```{r eval=FALSE, include=FALSE}
load(file = 'object/sgNTC/combined_seurat_filtered.Rdata')
##提取标准化表达矩阵
mat=as.data.frame(GetAssayData(cells.cut,layer = 'data'))%>%
	rownames_to_column(var = "rowname")%>%
	column_to_rownames(var = "rowname")
dim(mat);mat[1:10,1:10]
table(mat["Tcf7",]>2.5)

library(ggsci)
library(ggsignif)

if(T){
features=c("Tcf7","tox","btla","icam1",
		   "bcl6","cxcr5","pdcd1","tbx21")

# Step 1: 全小写
features <- tolower(features)
# Step 2: 首字母大写（任选一种方法）
# 方法一：基础R
features <- paste0(toupper(substr(features, 1, 1)), substr(features, 2, nchar(features)))
# 方法二：正则表达式
#features <- sub("^(.)", "\\U\\1", features, perl = TRUE)
# 方法三：stringr包
#features <- stringr::str_to_title(features)
features
}
mat_select=as.data.frame(t(mat[features,]))
str(mat_select)

colnames(cells.cut@meta.data)
# 检查行名是否一致
identical(rownames(cells.cut@meta.data), rownames(mat_select))
# 若一致，直接按列合并
merged_df <- cbind(cells.cut@meta.data, mat_select)

#小提琴图
colnames(merged_df)

wilcox_test_result_list=list()
dir.create("output/sgNTC/0018_geneexp_violin_dot_boxplot_chart")

cols=c("#ec5e6c","#6694cf","#cb7eb6","#f39139","#91c678","#f0e444")

for(i in features){
	#i=features[1]
df_meta = merged_df %>% 
	dplyr::select(assign_cluster, all_of(i))

# 执行两两比较的 Wilcoxon 秩和检验
wilcox_test_result <- pairwise.wilcox.test(df_meta[,i],
										   df_meta$assign_cluster, 
										   p.adjust.method = "bonferroni")
wilcox_test_result_list[[i]]=wilcox_test_result

write.csv(wilcox_test_result_list[[i]]$p.value,
		  file=paste("output/sgNTC/0018_geneexp_violin_dot_boxplot_chart/0018_",gsub("/", "", i),"_violin_dot_boxplot_chart_Pvalue.csv"),
		  na = "",row.names = T)

p1=merged_df%>%
ggplot(.,aes(x=assign_cluster,y= !!sym(i)))+
    # 添加小提琴图（先绘制，作为背景层）
  geom_violin(
    data = merged_df %>% dplyr::filter(!!sym(i) >= 0),  # 与箱线图相同的过滤条件
    aes(fill = assign_cluster),   # 使用相同填充色
    alpha = 1,                   # 调整透明度避免遮挡箱线图
    color = NA,                    # 移除小提琴边框线
    trim = TRUE,                   # 截断尾部以匹配数据范围
    scale = "width"                # 统一宽度
  ) +
	# 箱线图（覆盖在小提琴图上）
	geom_boxplot(data = merged_df %>% dplyr::filter(!!sym(i) > 0),  # 过滤条件
  	aes(fill = assign_cluster),  # 添加 fill 映射
  	 fill = NA,         # 强制不填充颜色[6](@ref)
  	notchwidth=1,
  	staplewidth=0,
    width = 0.3,
    color = "black",             # 箱线边框颜色（避免默认灰色覆盖填充色）
    size = 0.6,                  # 控制边缘线厚度（默认0.5，建议0.3-0.5）
    outliers=TRUE,
    outlier.color = "black",      # 异常点颜色（可选）
    outlier.size = 0.3  # 缩小异常点尺寸（默认2，建议设置为1-1.5）
    )+
	# 散点图（最后绘制，显示在顶层）
	 geom_jitter(
    aes(color = "black"),  # 关键修改：颜色映射到分组变量
    width = 0.4,
    alpha = 1,  # 调整透明度以增强可读性
    size = 0.1     # 可选：调整散点大小
    )+
  theme_classic()+xlab("")+ylab(paste0(i," exp."))+
	scale_fill_manual(values = cols) +
	scale_color_manual(values = "black") +
	theme(axis.text.x = element_text(size = 8,color = "black"),
  	  axis.text.y = element_text(size = 8, color = "black"),
  	  axis.title.x = element_text(size = 8, color = "black"),
  	  axis.title.y = element_text(size = 8, color = "black"),
  	  title = element_text(size = 8, color = "black"),
  	  legend.position = "none", # "none" REMOVE legend
		  )
p1

print(paste0("Now save gene plot of ",i))
pdf(paste("output/sgNTC/0018_geneexp_violin_dot_boxplot_chart/0018_",gsub("/", "", i),"_violin_dot_boxplot_chart.pdf", sep=""),width =5, height =3,compress=TRUE)
print(p1)
dev.off()

graph2ppt(p1,paste("output/sgNTC/0018_geneexp_violin_dot_boxplot_chart/0018_",
				gsub("/", "", i),"_violin_dot_boxplot_chart.pptx", sep=""),width =5, height =3)
}
qsavem(merged_df,
	   mat_select,
	   features,
	   wilcox_test_result_list,
	   cols,
	  file="object/sgNTC/0018_geneexp_violin_dot_boxplot_chart.qs")

qload("object/sgNTC/0018_geneexp_violin_dot_boxplot_chart.qs")
```

#pie chart of tfh ratio 
```{r eval=FALSE, include=FALSE}
load(file = 'object/sgNTC/combined_seurat_filtered.Rdata')

tab <- table(cells.cut$assign_cluster)
matrix_result <- data.frame(
  names = as.factor(names(tab)), 
  Count = as.numeric(as.vector(tab)),  # 显式数值化
  share = as.numeric(as.vector(tab))/sum(as.numeric(as.vector(tab)))*100
)

if(T){
color=c("#ec5e6c","#6694cf","#cb7eb6","#f39139","#91c678","#f0e444")
# 圆环图
pl=ggplot(matrix_result,aes(x = 1, y = Count, fill = names))+
  geom_bar(width = 0.3,stat = 'identity')+
  coord_polar(theta = 'y')+
  geom_text(aes(label = paste0(round(share,1),'%')),size=7,
            position = position_stack(vjust = .5))+
  labs(x = NULL, y = NULL, fill = NULL, 
       title = "")+
  scale_fill_manual(values = color) +
  theme_void() +
  theme(axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        plot.title = element_text(hjust = 0.5, color = "#666666"),
  	  legend.key.size = unit(0.75, "cm"),   # 图例图标大小（原默认0.8cm）
    legend.text = element_text(size=14),
      plot.margin = unit(rep(0,4), "cm"))+  # 图例文字字号（原默认11）
  xlim(0.5, 1.5)

pdf("output/sgNTC/0019_tfh_group_ratio_pie_chart.pdf", w=8, h=6)
print(pl)
dev.off()
}

save(matrix_result,
       file = "object/sgNTC/0019_tfh_group_ratio_pie_chart.Rdata")

load(file = "object/sgNTC/0019_tfh_group_ratio_pie_chart.Rdata")
```

#Stem-like & Effect./TD CD8 T signature
```{r eval=FALSE, include=FALSE}
suppressPackageStartupMessages({
	library("GenomicRanges")
	library("limma")
	library(edgeR)
	library(tidyverse)
	library("edgeR")
	library("DESeq2")
	library("clusterProfiler")
	library("org.Hs.eg.db")
	library("gplots")
	library(GEOquery)
	library(FactoMineR)#载入PCA函数
	library(factoextra)
})
gset<-getGEO(GEO ="GSE84105" ,GSEMatrix = TRUE,
			 AnnotGPL = TRUE,getGPL=F)

class(exprs(gset[[1]]))  # 应为 "matrix"
head(rownames(exprs(gset[[1]])))  # 应显示探针ID（如 "1000_at"）
pdata=pData(gset[[1]]);
exprs=as.data.frame(exprs(gset[[1]]));

library(biomaRt)
mart0 = useMart('ensembl')
listDatasets(mart0)
mart <- useMart("ensembl","mmusculus_gene_ensembl")  
#mmusculus_gene_ensembl	
#hsapiens_gene_ensembl
#查看biomaRt支持哪些ID类型的输入
input=listFilters(mart)#输入的ID类型
#查看biomaRt支持哪些ID类型的输出
output=listAttributes(mart)

gene=rownames(exprs)

geneID <- getBM(attributes=
                  c("external_gene_name","affy_mouse430_2"),
                filters ="affy_mouse430_2",values = gene, mart = mart)#[,1]
geneID # 查看
# filters，指定输入ID的类型，
# attributes指定输出ID的类型，
# [,1]，取第一列，是为了把输出的ID从矩阵变为向量，因为clusterProfiler包的enrichKEGG函数需要输入为向量。

exprs$affy_mouse430_2=rownames(exprs)
mRNA<-dplyr::filter(geneID)%>%#选择编码蛋白
  dplyr::select(external_gene_name,affy_mouse430_2)%>%
  inner_join(exprs,by ="affy_mouse430_2")%>%#与表达谱合并
  dplyr::distinct(external_gene_name,.keep_all = T)
row.names(mRNA)=mRNA$external_gene_name;mRNA=mRNA[,-c(1:2)]

#CXCR5+_CD8 Stem-like cell CD8
#Tim3+_CD8  Effect./TD CD8

# 判断数据质量 : PCA进行分组查看
colnames(mRNA)=c("Naïve_1","Naïve_2",
		 "Stem-like_1","Stem-like_2","Stem-like_3",
		 "Effect/TD_1","Effect/TD_2","Effect/TD_3")
if(T){
### 判断数据质量 : PCA进行分组查看
group4=c("Naïve","Naïve",
		 "Stem-like","Stem-like","Stem-like",
		 "Effect/TD","Effect/TD","Effect/TD")

group_list=factor(group4,levels = c("Naïve","Stem-like","Effect/TD"))
### 筛选MAD前5000的基因(MAD（绝对中位差）反映数据的变化程度)
keep_data <- mRNA[order(apply(mRNA,1,mad), decreasing = T)[1:5000],]
dat.pca <- PCA(t(keep_data), graph = F) 
pca <- fviz_pca_ind(dat.pca,axes = c(1, 2),
                    title = "Principal Component Analysis",
                    legend.title = "Groups",
                    geom.ind = c("point","text"), #"point","text"
                    pointsize = 3,
                    labelsize = 4,
                    repel = TRUE, #标签不重叠
                    col.ind = group_list, # 分组上色
                    axes.linetype=NA,  # remove axeslines
                    mean.point=F#去除分组中心点
                    ) +
  theme(axis.text.x = element_text(size = 8,color = "black"),
  	  axis.text.y = element_text(size = 8, color = "black"),
  	  axis.title.x = element_text(size = 14, color = "black"),
  	  axis.title.y = element_text(size = 14, color = "black"),
  	  title = element_text(size = 14, color = "black", face = "bold"),
  	  legend.position = "none", # "none" REMOVE legend
        plot.title = element_text(hjust = 0.5),
  	  panel.border = element_rect(colour = "black", fill = NA, linewidth = 1), #加外框线
  	  panel.grid.major = element_blank(),  # 移除主网格线
    panel.grid.minor = element_blank(),  # 移除次网格线
    panel.background = element_blank()   # 移除背景填充
  )+  
  coord_fixed(ratio = 1) #坐标轴的纵横比
pca
ggsave(pca,filename= "output/sgNTC/0020_Stem-like_EffectTD_CD8T_PCA_analysis.pdf", width = 5, height = 5)
}
```

#limma DEG Stem-like_vs_Naïve
```{r eval=FALSE, include=FALSE}
if(T){
group4=c("Naïve","Naïve",
		 "Stem-like","Stem-like","Stem-like")

group_list=factor(group4,levels = c("Naïve","Stem-like"))
table(group_list)#检查一下组别数量

}

if(T){
  N_counts=mRNA[,c(1:5)]
  y <- DGEList(counts=N_counts)
  keep <- rowSums(cpm(y)>1) >= 2
y <- y[keep, , keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
logCPM <- cpm(y, log=TRUE, prior.count=3)
design <- model.matrix(~group_list)
fit <- lmFit(logCPM ,design)
fit1 <- eBayes(fit,trend=TRUE)
summary(decideTests(fit))

DEG_limma = topTable(fit1, coef=2, n=Inf)
DEG_limma = na.omit(DEG_limma)
pvalue_t=0.05
logFC_t=1
k1 = (DEG_limma$adj.P.Val < pvalue_t)&(DEG_limma$logFC < -logFC_t)
k2 = (DEG_limma$adj.P.Val < pvalue_t)&(DEG_limma$logFC > logFC_t)
DEG_limma$regulated = ifelse(k1,"down",ifelse(k2,"up","not"))
table(DEG_limma$regulated)

head(DEG_limma[DEG_limma$regulated=="up",])

up=head(DEG_limma[DEG_limma$regulated=="up",])[1,] #直接选取第一列上调基因
upgene <- rownames(up)
topexp <- c(t(N_counts[match(upgene,rownames(N_counts)),]))
topexp

test <- data.frame(value=topexp, group=group_list)
test

write.csv(DEG_limma,file = "output/sgNTC/0020_Stem-like_vs_Naïve_DEGlimma.csv",row.names = T)
}

save(N_counts,gset,pdata,exprs,mRNA,
	 DEG_limma,
	 file = 'object/sgNTC/0020_Stem-like_vs_Naïve_DEGlimma.Rdata')

load(file = 'object/sgNTC/0020_Stem-like_vs_Naïve_DEGlimma.Rdata')
```

#limma DEG Effect/TD_vs_Naïve
```{r eval=FALSE, include=FALSE}
if(T){
group4=c("Naïve","Naïve",
		 "Effect/TD","Effect/TD","Effect/TD")

group_list=factor(group4,levels = c("Naïve","Effect/TD"))
table(group_list)#检查一下组别数量

}

if(T){
  N_counts=mRNA[,c(1:2,6:8)]
  y <- DGEList(counts=N_counts)
  keep <- rowSums(cpm(y)>1) >= 2
y <- y[keep, , keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
logCPM <- cpm(y, log=TRUE, prior.count=3)
design <- model.matrix(~group_list)
fit <- lmFit(logCPM ,design)
fit1 <- eBayes(fit,trend=TRUE)
summary(decideTests(fit))

DEG_limma = topTable(fit1, coef=2, n=Inf)
DEG_limma = na.omit(DEG_limma)
pvalue_t=0.05
logFC_t=1
k1 = (DEG_limma$adj.P.Val < pvalue_t)&(DEG_limma$logFC < -logFC_t)
k2 = (DEG_limma$adj.P.Val < pvalue_t)&(DEG_limma$logFC > logFC_t)
DEG_limma$regulated = ifelse(k1,"down",ifelse(k2,"up","not"))
table(DEG_limma$regulated)

head(DEG_limma[DEG_limma$regulated=="up",])

up=head(DEG_limma[DEG_limma$regulated=="up",])[1,] #直接选取第一列上调基因
upgene <- rownames(up)
topexp <- c(t(N_counts[match(upgene,rownames(N_counts)),]))
topexp

test <- data.frame(value=topexp, group=group_list)
test

write.csv(DEG_limma,file = "output/sgNTC/0020_EffectTD_vs_Naïve_DEGlimma.csv",row.names = T)
}

save(N_counts,gset,pdata,exprs,mRNA,
	 DEG_limma,
	 file = 'object/sgNTC/0020_EffectTD_vs_Naïve_DEGlimma.Rdata')

load(file = 'object/sgNTC/0020_EffectTD_vs_Naïve_DEGlimma.Rdata')
```

#Creat custom gene set of Stem-like_Effect/TD differentiation 
```{r eval=FALSE, include=FALSE}
rm(list =ls());gc()
suppressPackageStartupMessages({
	library(clusterProfiler)
	library(tidyverse)
	library(ggplot2)
	library(org.Hs.eg.db)
	library(org.Mm.eg.db)
	library(GSEABase)
})

if(T){
load(file = 'object/sgNTC/0020_Stem-like_vs_Naïve_DEGlimma.Rdata')
Stemlike_vs_Naïve=DEG_limma

load(file = 'object/sgNTC/0020_EffectTD_vs_Naïve_DEGlimma.Rdata')
EffectTD_vs_Naïve=DEG_limma

Stemlike_differentiation_signature=Stemlike_vs_Naïve%>%
  arrange(desc(logFC))%>%
  top_n(n=300,wt=logFC)

EffectTD_differentiation_signature=EffectTD_vs_Naïve%>%
  arrange(desc(logFC))%>%
  top_n(n=300,wt=logFC)
}
mRNA["Ahr",]
Stemlike=rownames(Stemlike_differentiation_signature)%in%
        rownames(EffectTD_differentiation_signature)
Stemlike_df=Stemlike_differentiation_signature[!Stemlike,]

EffectTD=rownames(EffectTD_differentiation_signature)%in%
	rownames(Stemlike_differentiation_signature)
EffectTD_df=EffectTD_differentiation_signature[!EffectTD,]

if(T){
pathways=list()
pathways[["Stem-like cell CD8 T cell gene signature"]]=
	rownames(Stemlike_differentiation_signature)[!Stemlike]
pathways[["Effect/TD CD8 T cell gene signature"]]=
	rownames(EffectTD_differentiation_signature)[!EffectTD]

# 创建一个 GMT 文件
gmt_file_path <- "input/gmt/Stemlike_EffectTD_differentiation_signature.gmt"

# 构造数据帧
data_rows <- lapply(names(pathways), function(name) {
  genes <- pathways[[name]]
  data.frame(
    GeneSetName = name,
    GeneSetDescription = name,
    gene_names = genes,
    stringsAsFactors = FALSE)
})

# 合并所有行
data_df <- do.call(rbind, data_rows)

# 写入 GMT 文件
write.table(data_df,
            file = gmt_file_path,
            row.names = FALSE,
            col.names = FALSE,
            sep = "\t",
            quote = FALSE
)
}
  d="input/gmt/Stemlike_EffectTD_differentiation_signature.gmt"
geneset <- read.gmt(d)

write.csv(geneset, 
            file ="output/sgNTC/0020_Stemlike_EffectTD_differentiation_signature.csv",row.names = F)

write.csv(Stemlike_df, 
            file ="output/sgNTC/0020_Stemlike_signature_df.csv",row.names = T)
write.csv(EffectTD_df, 
            file ="output/sgNTC/0020_EffectTD_signature_df.csv",row.names = T)

save(data_df,
	 Stemlike_vs_Naïve,EffectTD_vs_Naïve,
	 Stemlike_differentiation_signature,EffectTD_differentiation_signature,
	 pathways,
	 EffectTD,Stemlike,
	 EffectTD_df,Stemlike_df,
	 file = 'object/sgNTC/0020_Stemlike_EffectTD_differentiation_signature.Rdata')

load(file = 'object/sgNTC/0020_Stemlike_EffectTD_differentiation_signature.Rdata')
```


#Running GSVA with Custom Gene Sets

```{r eval=FALSE, include=FALSE}
suppressPackageStartupMessages({
	library(GSVA)
library(GSEABase)
library(tidyverse)
library(ggplot2)
#install.packages("PupillometryR")
library(PupillometryR)
library(ggforce)
library(dplyr) 
})

load(file = 'object/sgNTC/combined_seurat_filtered.Rdata')

geneset <- read.gmt(gmtfile="input/gmt/Stemlike_EffectTD_differentiation_signature.gmt")
head(geneset)

if(T){
# 按 term 列分组并收集基因
geneset_list <- geneset %>%
  group_by(term) %>%
  summarise(genes = list(gene), .groups = 'drop') %>%
  pull(genes) %>%
  set_names(unique(geneset$term))


gene.expr <-  as.matrix(cells.cut[["RNA"]]$scale.data)
dim(gene.expr)
gsva.result <- GSVA::gsva(gene.expr, 
						  gset.idx.list=geneset_list,
						  kcdf='Gaussian')

gsva_long <- reshape2::melt(gsva.result)%>%
	dplyr::rename(pathway_name="Var1",
		   cell_id="Var2",
		   pathway_score="value")
scMetadata=cells.cut@meta.data%>%
	rownames_to_column(var = "cell_id")

gsva_long_meta=merge(gsva_long,scMetadata,
				by.x ="cell_id", by.y = "cell_id", all = TRUE)

names(gsva_long_meta);min(gsva_long_meta$pathway_score);max(gsva_long_meta$pathway_score)

wilcox_test_result_list=list()
for(i in levels(gsva_long_meta$pathway_name)){
gsva_long_meta01=gsva_long_meta%>%
	dplyr::filter(pathway_name%in% i)

# 执行两两比较的 Wilcoxon 秩和检验
wilcox_test_result <- pairwise.wilcox.test(gsva_long_meta01$pathway_score,
										   gsva_long_meta01$assign_cluster, 
										   p.adjust.method = "bonferroni")
wilcox_test_result_list[[i]]=wilcox_test_result

p1 <- 
    ggplot(data = gsva_long_meta01, 
           aes(x = assign_cluster, y = pathway_score, fill = assign_cluster)) +
    geom_flat_violin(position = position_nudge(x = 0.2, y = 0), alpha = 0.8) +
    geom_point(aes(y = pathway_score, color = assign_cluster), 
               position = position_jitter(width = 0.15), size = 1, alpha = 0.1) +
    geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.8) +
    labs(y = NULL, x = NULL) +  #Normalized z-score
    guides(fill = "none", color = "none") +
    scale_y_continuous(limits = c(-0.6, 0.6)) +
    scale_fill_manual(values = c("#ec5e6c","#6694cf","#cb7eb6","#f39139","#91c678","#f0e444")) +
    scale_colour_manual(values = c("#ec5e6c","#6694cf","#cb7eb6","#f39139","#91c678","#f0e444")) +
    coord_flip() +	
	labs(title=paste0(i,"\n"))+
	theme_classic()+
	theme(axis.text.x = element_text(size = 14, color = "black", family = "sans"),
  	  axis.text.y = element_text(size = 14, color = "black", family = "sans"),
		  axis.title = element_text(size = 16),
		  plot.title = element_text(hjust = 0.5, size = 16))

pdf(paste("output/sgNTC/0020_",gsub("/", "", i),"gsva_result.pdf", sep=""),width =6, height =4,compress=TRUE)
print(p1)
dev.off()
}

save(gsva.result,
	 geneset_list,
	 scMetadata,
	 wilcox_test_result_list,
	 gsva_long_meta,
	 file = 'object/sgNTC/0020_Stemlike_EffectTD_differentiation_signature_gsva_result.Rdata')
}

rm(list =ls())
load(file = 'object/sgNTC/0020_Stemlike_EffectTD_differentiation_signature_gsva_result.Rdata')
```

#C1-C3 reanalysis
```{r eval=FALSE, include=FALSE}
rm(list =ls());gc()

load(file = 'object/sgNTC/combined_seurat_filtered.Rdata')

cellscut_c1_c3 <- subset(cells.cut, subset = assign_cluster=="c1" | 
					 	assign_cluster=="c2" | 
					 	assign_cluster=="c3")
cellscut_c1_c3$assign_cluster=factor(
	as.character(cellscut_c1_c3$assign_cluster))
table(cellscut_c1_c3$assign_cluster)

cellscut_c1_c3 <- NormalizeData(cellscut_c1_c3 ,
						   normalization.method = "LogNormalize", scale.factor = 10000)
#高变基因
cellscut_c1_c3 <- FindVariableFeatures(cellscut_c1_c3, selection.method = "vst", nfeatures = 3000)
gene_data <- VariableFeatures(object =cellscut_c1_c3)

#归一化
cellscut_c1_c3 <- ScaleData(cellscut_c1_c3, features = gene_data)
#降维聚类
cellscut_c1_c3 <- RunPCA(cellscut_c1_c3, 
					features = VariableFeatures(object = cellscut_c1_c3))
ElbowPlot(cellscut_c1_c3)
cellscut_c1_c3 <- FindNeighbors(cellscut_c1_c3, dims = 1:10)
cellscut_c1_c3 <- FindClusters(cellscut_c1_c3, resolution = 0.5)
cellscut_c1_c3 <- RunUMAP(cellscut_c1_c3, dims = 1:3,
					 n.components=2,
					 metric = "manhattan", 
					 #euclidean, *****
					 #cosine, ***
					 #manhattan, *****
					 #hamming, 
					 #correlation, ***
					 #precomputed
					 umap.method="umap-learn",return.model = TRUE,
					 min.dist=0.2,
					 n.neighbors=50L)

cellscut_c1_c3@active.ident=cellscut_c1_c3$assign_cluster
cellscut_c1_c3$seurat_clusters=cellscut_c1_c3@active.ident
table(cellscut_c1_c3$seurat_clusters)
```

#UMAP plot
```{r eval=FALSE, include=FALSE}
cellscut_c1_c3@active.ident=as.factor(cellscut_c1_c3$assign_cluster)
color=c("#ec5e6c","#6694cf","#cb7eb6","#f39139","#91c678","#f0e444")

if(T){
p3 <- DimPlot(cellscut_c1_c3, reduction = "umap", 
              group.by = "assign_cluster", pt.size=0.8,
              cols = color,label = TRUE,
              label.size = 5,label.color = "black",
			  order = TRUE,
              label.box = FALSE,raster=FALSE)+
	theme(axis.line = element_blank(),
		  axis.ticks = element_blank(),
		  axis.text = element_blank(),
		  panel.grid = element_blank(),
		  legend.position = "none")+
	xlab(NULL)+ylab(NULL)+	
	labs(title=NULL)

pdf(file = "output/sgNTC/0021_C1-C3_merged_UMAP.pdf",width = 5, height=4,compress = TRUE)
print(p3)
dev.off()
}
if(T){
table(cellscut_c1_c3$orig.ident)
p3 <- DimPlot(cellscut_c1_c3, reduction = "umap", 
              group.by = "orig.ident", pt.size=0.8,
              cols = color,label = FALSE,
              label.size = 5,label.color = "black",
              label.box = FALSE,raster=FALSE)+
	theme(axis.line = element_blank(),
		  axis.ticks = element_blank(),
		  axis.text = element_blank(),
		  panel.grid = element_blank(),
		  legend.position = "none")+
	xlab(NULL)+ylab(NULL)+	
	labs(title=NULL)+
  facet_wrap(~ orig.ident, ncol = 1)

pdf(file = "output/sgNTC/0021_C1-C3_merged_UMAP_origident.pdf",width = 3, height=6,compress = TRUE)
print(p3)
dev.off()
}

pbmc.markers <- FindAllMarkers(cellscut_c1_c3,test.use = "wilcox",
							   only.pos = F,min.pct = 0,logfc.threshold = 0,
							   return.thresh = 1)
write.csv(pbmc.markers, 
            file ="output/sgNTC/0021_sgNTC_C1-C3_DEG_result.csv")

save(cellscut_c1_c3,
	 pbmc.markers,
	 file = 'object/sgNTC/0021_sgNTC_C1-C3_combined_seurat_filtered.Rdata')

rm(list =ls());gc()
load(file = 'object/sgNTC/0021_sgNTC_C1-C3_combined_seurat_filtered.Rdata')
```
